esphome/components/bme68x_bsec2_i2c/__init__.py

```diff
@@ -0,0 +1,226 @@
+
+AlgorithmOutput = bme68x_bsec2_i2c_ns.enum("AlgorithmOutput")
+ALGORITHM_OUTPUT_OPTIONS = {
+    "classification": AlgorithmOutput.ALGORITHM_OUTPUT_CLASSIFICATION,
+    "regression": AlgorithmOutput.ALGORITHM_OUTPUT_REGRESSION,
+}
+
+OperatingAge = bme68x_bsec2_i2c_ns.enum("OperatingAge")
+OPERATING_AGE_OPTIONS = {
+    "4d": OperatingAge.OPERATING_AGE_4D,
+    "28d": OperatingAge.OPERATING_AGE_28D,
+}
+
+SampleRate = bme68x_bsec2_i2c_ns.enum("SampleRate")
+SAMPLE_RATE_OPTIONS = {
+    "LP": SampleRate.SAMPLE_RATE_LP,
+    "ULP": SampleRate.SAMPLE_RATE_ULP,
+}
+
+Voltage = bme68x_bsec2_i2c_ns.enum("Voltage")
+VOLTAGE_OPTIONS = {
+    "1.8V": Voltage.VOLTAGE_1_8V,
+    "3.3V": Voltage.VOLTAGE_3_3V,
+}
+
+BME68xBSEC2I2CComponent = bme68x_bsec2_i2c_ns.class_(
+    "BME68xBSEC2I2CComponent", cg.Component, i2c.I2CDevice
+)
+
+ALGORITHM_OUTPUT_FILE_NAME = {
+    "classification": "sel",
+    "regression": "reg",
+}
+
+SAMPLE_RATE_FILE_NAME = {
+    "LP": "3s",
+    "ULP": "300s",
+}
+
+VOLTAGE_FILE_NAME = {
+    "1.8V": "18v",
+    "3.3V": "33v",
+}
+
+
+def _compute_local_file_path(url: str) -> Path:
+    h = hashlib.new("sha256")
+    h.update(url.encode())
+    key = h.hexdigest()[:8]
+    base_dir = external_files.compute_local_file_dir(DOMAIN)
+    return base_dir / key
+
+
+def _compute_url(config: dict) -> str:
+    model = config.get(CONF_MODEL)
+    operating_age = config.get(CONF_OPERATING_AGE)
+    sample_rate = SAMPLE_RATE_FILE_NAME[config.get(CONF_SAMPLE_RATE)]
+    volts = VOLTAGE_FILE_NAME[config.get(CONF_SUPPLY_VOLTAGE)]
+    if model == "bme688":
+        algo = ALGORITHM_OUTPUT_FILE_NAME[
+            config.get(CONF_ALGORITHM_OUTPUT, "classification")
+        ]
+        filename = "bsec_selectivity"
+    else:
+        algo = "iaq"
+        filename = "bsec_iaq"
+    return f"https://raw.githubusercontent.com/boschsensortec/Bosch-BSEC2-Library/{BSEC2_LIBRARY_VERSION}/src/config/{model}/{model}_{algo}_{volts}_{sample_rate}_{operating_age}/{filename}.txt"
+
+
+def _download_file(url: str, path: Path) -> bytes:
+    if not external_files.has_remote_file_changed(url, path):
+        _LOGGER.debug("Remote file has not changed, skipping download")
+        return path.read_bytes()
+
+    try:
+        req = requests.get(
+            url,
+            timeout=external_files.NETWORK_TIMEOUT,
+            headers={"User-agent": f"ESPHome/{__version__} (https://esphome.io)"},
+        )
+        req.raise_for_status()
+    except requests.exceptions.RequestException as e:
+        raise cv.Invalid(f"Could not download file from {url}: {e}") from e
+
+    path.parent.mkdir(parents=True, exist_ok=True)
+    path.write_bytes(req.content)
+    return req.content
+
+
+def download_bme68x_blob(config):
+    url = _compute_url(config)
+    path = _compute_local_file_path(url)
+
+    if len(_download_file(url, path)) > 0:
+        return config
+
+    raise cv.Invalid("Unable to download binary configuration blob")
```

Comment:
This should simplify a bit
The download content will raise cv.Invalid if it failed
```suggestion
def download_bme68x_blob(config):
    url = _compute_url(config)
    path = _compute_local_file_path(url)

    external_files.download_content(url, path):
    return config
```

Reply:
Oh sweet! 😄 Thank you